pymc4/model.py

```diff
@@ -0,0 +1,126 @@
+import tensorflow as tf
+import seaborn as sns
+import threading
+import matplotlib.pyplot as plt
+
+
+class Context(object):
+    """Functionality for objects that put themselves in a context using
+    the `with` statement.
+    """
+    contexts = threading.local()
+
+    def __enter__(self):
+        type(self).get_contexts().append(self)
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        type(self).get_contexts().pop()
+
+
+def withparent(meth):
+    """Helper wrapper that passes calls to parent's instance"""
+    def wrapped(self, *args, **kwargs):
+        res = meth(self, *args, **kwargs)
+        if getattr(self, 'parent', None) is not None:
+            getattr(self.parent, meth.__name__)(*args, **kwargs)
+        return res
+    # Unfortunately functools wrapper fails
```

Comment:
i am not sure this is true any more - we should have a test that fails without `withparent`, or else get rid of it

Reply:
I am planning to remove all dependency treedict. This shouldn't be even required once I do that.